lib/path-design.md

```diff
@@ -0,0 +1,373 @@
+
+# normalise the path
+relative.join ["./foo" "bar//./baz/" "./qux" ] == "foo/bar/baz/qux"
+
+# empty list is the current directory
+relative.join [] == "."
+
+# do not handle parent directory, as it may break underlying symlinks
+relative.join ["foo" ".."] == <error>
+
+# do not handle absolute paths elements
+relative.join ["/foo" "bar"] == <error>
+relative.join ["foo" "/bar"] == <error>
+relative.join ["foo" "/" ] == <error>
+```
+
+Laws:
+- Associativity:
+  `relative.join [ x (join [ y z ]) ] == relative.join [ (join [ x y ]) z ]`
+- Identity:
+  `relative.join [] == "."`
+  `relative.join [p "."] == normalise p`
+  `relative.join ["." p] == normalise p`
+- The result is normalised:
+  `relative.join ps == normalise (relative.join ps)`
+- Joining a single path is that path itself, but normalised:
+  `relative.join [ p ] == normalise p`
+
+### `relative.normalise`
+
+```haskell
+relative.normalise :: String -> String
+```
+
+Normalise relative paths.
+
+- Limit repeating `/` to a single one (does not change a [POSIX Pathname](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_271))
+- Remove redundant `.` components (does not change the result of [POSIX Pathname Resolution](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_13))
+- Error on empty strings (not allowed as a [POSIX Filename](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_170))
+- Remove trailing `/` and `/.` (See [justification](#trailing-slashes))
+- Error on `..` path components (See [justification](#parents))
+- Remove leading `./` (See [justification](#leading-dots))
+
+Examples:
+```
+# limit repeating `/` to a single one
+relative.normalise "foo//bar" == "foo/bar"
+
+# remove redundant `.` components
+relative.normalise "foo/./bar" == "foo/bar"
+
+# remove leading `./`
+# TODO: bikeshedding. there is a use to this.
+relative.normalise "./foo/bar" == "foo/bar"
+
+# remove trailing `/`
+relative.normalise "foo/bar/" == "foo/bar"
+
+# remove trailing `/.`
+relative.normalise "foo/bar/." == "foo/bar"
+
+# error on `..` path components
+relative.normalise "foo/../bar" == <error>
+
+# error on empty string
+relative.normalise "" == <error>
+
+# error on absolute path
+relative.normalise "/foo" == <error>
+```
+
+Laws:
+- Idempotency:
+
+  `relative.normalise (relative.normalise p) == relative.normalise p`
+
+- Doesn't change the file system object pointed to:
+
+  `$(stat ${p}) == $(stat ${relative.normalise p})`
+
+  TODO: Is this the same as the law below?
+
+- Uniqueness: If the normalisation of two paths is different then they point to different paths:
+
+  `normalise p != normalise q => $(stat ${p}) != $(stat ${q})`
+
+  Note: This law only holds if trailing slashes are not persisted, see the [trailing slashes decision](#trailing-slashes)
+
+### `difference`
+
+```haskell
+difference :: AttrsOf Path -> { commonPrefix :: Path; suffix :: AttrsOf String; }
```

Comment:
Ideas:

```
commonAncestry :: f Path -> { commonAncestor :: Path; suffix :: f String; }
  where f is one of
    attrsOf
    listOf
```

Reply:
Since a path is a sequence of nodes in a tree, the correct term for the object of interest is "prefix", not "ancestor". The latter applies to the relations of nodes, not elements in a path.

I'm not the biggest fan of taking a list. With @infinisil we crystallised a few key ideas about the library, also based on your feedback, and one of them was that it should be as general as possible such that one can implement utility functions on top of it. A list has strictly less information than an attrset, and can always be projected down from the attrset, but not the other way around.

```haskell
commonPrefix :: AttrsOf Path -> { prefix :: Path; suffix :: AttrsOf String }
```

```nix
{ 
commonPrefixList xs =
  let
    intermediate = commonPrefix (listToAttrs (imap0 (i: x: { ${toString i} = x; })));
  in { prefix = intermediate.prefix; suffix = attrValues intermediate.suffix; };
```

Sure that's slow, but right now I don't even know if we have a use case for that interface.